lib/morandi/image_ops.rb

```diff
@@ -315,4 +315,56 @@ def call(_image, pixbuf)
       end
     end
   end
+
+  class ShrinkToFit < ImageOp
+    attr_accessor :crop, :size, :print_size, :shrink
+
+    def call(_image, pixbuf)
+      return unless shrink
+
+      original_img_height = pixbuf.height # Original size
+      original_img_width = pixbuf.width # Original size
+
+      shrink_x, shrink_y, _width, _height = crop
+      output_width, output_height = print_size
+
+      surface = Cairo::ImageSurface.new(:rgb24, output_width, output_height)
+      cr = Cairo::Context.new(surface)
+
+      # Apply White background
+      cr.save do
+        cr.set_operator :source
+        cr.set_source_rgb 1, 1, 1
+        cr.paint
+        cr.fill
+      end
+      # end
+
+      cr.translate(shrink_x.abs, shrink_y.abs)
+      scale_by = largest_shrink_ratio(output_width, output_height, original_img_width, original_img_height)
+      cr.scale(scale_by, scale_by)
+      cr.set_source_pixbuf(pixbuf)
+
+      cr.paint(1.0)
+      final_pb = surface.to_pixbuf
+
+      # Clean up
+      cr.destroy
+      surface.destroy
+      final_pb
+    end
+
+    private
+
+    def largest_shrink_ratio(print_width, print_height, image_width, image_height)
+      proportional_width = print_width / image_width.to_f
+      proportional_height = print_height / image_height.to_f
+
+      return 1 if proportional_height >= 1 && proportional_width >= 1
+
+      options = { proportional_width => [(image_width * proportional_width).round, (image_height * proportional_width).round],
+                  proportional_height => [(image_width * proportional_height).round, (image_height * proportional_height).round] }
+      options.key(options.values.select { |option| option[0] <= print_width && option[1] <= print_height }.max)
+    end
```

Comment:
I'm glad you understand what's going on :laughing: Perhaps using a Hash as the data structure here helps the computer more than me.

Does anything here have significance, like "is this a square shape, so act like this" if you see what I mean? Line 363 seems most likely to have a meaning I'm not picking up on.

Reply:
The hash was honestly the easiest means of returning which number to use as scaling with Cairo - I'd originally used a two dimensional array so it came out with a calculated height and width but that doesn't actually work with how Cairo does things. If the image is square then the hash will have two identical keys (and scaled lengths/widths) on creation, as hashes must have unique keys it will keep the second entry and discard the first - It will still work as expected, just only have one option to pick from instead of two on line 367

Line 363 escapes the method if it would be expanding the image to fit the print size, ie if the picture is smaller than the print it will not blow it up and the picture will just scale itself by 1 (remain the same size)

Reply:
Ah, it's got a multipler, so "shrinking" can't get bigger than the size it already is! Gotcha :)